{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1f7b69bd-9eca-4b63-beea-ffdc5de11cba",
   "metadata": {},
   "source": [
    "Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "Note: Here Exception class refers to the base class for all the exceptions.\n",
    "\n",
    "\n",
    "\n",
    "When creating a custom exception in Python, it is recommended to inherit from the base Exception class. The Exception class provides a set of attributes and methods that are commonly used in exception handling, such as the ability to set an error message and retrieve the traceback information.\n",
    "\n",
    "By inheriting from the Exception class, the custom exception can leverage the functionality provided by the base class, such as the ability to be caught by a general exception handler that catches all exceptions, or by a more specific handler that only catches exceptions of a particular type.\n",
    "\n",
    "Additionally, the Exception class ensures that the custom exception conforms to the Python exception hierarchy, which helps maintain consistency and clarity in the code. It also ensures that the custom exception can be used in conjunction with other built-in exception classes in Python.\n",
    "\n",
    "Therefore, by inheriting from the Exception class, a custom exception gains access to a wide range of built-in exception handling functionality and ensures that the custom exception is compatible with the Python exception hierarchy."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbe6a0f3-d37c-4e7e-a279-f6a46fda0be4",
   "metadata": {},
   "source": [
    "Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "28583257-6ef8-4104-9146-4573446b015e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TypeError\n",
      "  |- FloatOperation\n",
      "  |- MultipartConversionError\n",
      "StopAsyncIteration\n",
      "StopIteration\n",
      "ImportError\n",
      "  |- ModuleNotFoundError\n",
      "  |- ZipImportError\n",
      "OSError\n",
      "  |- ConnectionError\n",
      "  |- BlockingIOError\n",
      "  |- ChildProcessError\n",
      "  |- FileExistsError\n",
      "  |- FileNotFoundError\n",
      "  |- IsADirectoryError\n",
      "  |- NotADirectoryError\n",
      "  |- InterruptedError\n",
      "  |- PermissionError\n",
      "  |- ProcessLookupError\n",
      "  |- TimeoutError\n",
      "  |- UnsupportedOperation\n",
      "  |- itimer_error\n",
      "  |- herror\n",
      "  |- gaierror\n",
      "  |- SSLError\n",
      "  |- Error\n",
      "  |- SpecialFileError\n",
      "  |- ExecError\n",
      "  |- ReadError\n",
      "  |- URLError\n",
      "  |- BadGzipFile\n",
      "EOFError\n",
      "  |- IncompleteReadError\n",
      "RuntimeError\n",
      "  |- RecursionError\n",
      "  |- NotImplementedError\n",
      "  |- _DeadlockError\n",
      "  |- BrokenBarrierError\n",
      "  |- BrokenExecutor\n",
      "  |- SendfileNotAvailableError\n",
      "  |- ExtractionError\n",
      "  |- VariableError\n",
      "NameError\n",
      "  |- UnboundLocalError\n",
      "AttributeError\n",
      "  |- FrozenInstanceError\n",
      "SyntaxError\n",
      "  |- IndentationError\n",
      "LookupError\n",
      "  |- IndexError\n",
      "  |- KeyError\n",
      "  |- CodecRegistryError\n",
      "ValueError\n",
      "  |- UnicodeError\n",
      "  |- UnsupportedOperation\n",
      "  |- JSONDecodeError\n",
      "  |- SSLCertVerificationError\n",
      "  |- Error\n",
      "  |- UnsupportedDigestmodError\n",
      "  |- IllegalMonthError\n",
      "  |- IllegalWeekdayError\n",
      "  |- ParserError\n",
      "  |- ClassNotFound\n",
      "  |- ClipboardEmpty\n",
      "  |- MessageDefect\n",
      "  |- MacroToEdit\n",
      "  |- InvalidFileException\n",
      "  |- UnequalIterablesError\n",
      "  |- InvalidVersion\n",
      "  |- _InvalidELFFileHeader\n",
      "  |- InvalidWheelFilename\n",
      "  |- InvalidSdistFilename\n",
      "  |- InvalidSpecifier\n",
      "  |- InvalidMarker\n",
      "  |- UndefinedComparison\n",
      "  |- UndefinedEnvironmentName\n",
      "  |- InvalidRequirement\n",
      "  |- InvalidVersion\n",
      "AssertionError\n",
      "ArithmeticError\n",
      "  |- FloatingPointError\n",
      "  |- OverflowError\n",
      "  |- ZeroDivisionError\n",
      "  |- DecimalException\n",
      "SystemError\n",
      "  |- CodecRegistryError\n",
      "ReferenceError\n",
      "MemoryError\n",
      "BufferError\n",
      "Warning\n",
      "  |- UserWarning\n",
      "  |- EncodingWarning\n",
      "  |- DeprecationWarning\n",
      "  |- PendingDeprecationWarning\n",
      "  |- SyntaxWarning\n",
      "  |- RuntimeWarning\n",
      "  |- FutureWarning\n",
      "  |- ImportWarning\n",
      "  |- UnicodeWarning\n",
      "  |- BytesWarning\n",
      "  |- ResourceWarning\n",
      "  |- DeprecatedTzFormatWarning\n",
      "  |- PkgResourcesDeprecationWarning\n",
      "_OptionError\n",
      "_Error\n",
      "error\n",
      "Verbose\n",
      "Error\n",
      "SubprocessError\n",
      "  |- CalledProcessError\n",
      "  |- TimeoutExpired\n",
      "TokenError\n",
      "StopTokenizing\n",
      "ClassFoundException\n",
      "EndOfBlock\n",
      "TraitError\n",
      "Error\n",
      "Error\n",
      "  |- CancelledError\n",
      "  |- TimeoutError\n",
      "  |- InvalidStateError\n",
      "_GiveupOnSendfile\n",
      "error\n",
      "Incomplete\n",
      "TimeoutError\n",
      "InvalidStateError\n",
      "LimitOverrunError\n",
      "QueueEmpty\n",
      "QueueFull\n",
      "Empty\n",
      "Full\n",
      "ArgumentError\n",
      "ZMQBaseError\n",
      "  |- ZMQError\n",
      "  |- ZMQBindError\n",
      "  |- NotDone\n",
      "PickleError\n",
      "  |- PicklingError\n",
      "  |- UnpicklingError\n",
      "_Stop\n",
      "ArgumentError\n",
      "ArgumentTypeError\n",
      "ConfigError\n",
      "  |- ConfigLoaderError\n",
      "  |- ConfigFileNotFound\n",
      "ConfigurableError\n",
      "  |- MultipleInstanceError\n",
      "ApplicationError\n",
      "error\n",
      "TimeoutError\n",
      "error\n",
      "ReturnValueIgnoredError\n",
      "KeyReuseError\n",
      "UnknownKeyError\n",
      "LeakedCallbackError\n",
      "BadYieldError\n",
      "ReturnValueIgnoredError\n",
      "Return\n",
      "InvalidPortNumber\n",
      "error\n",
      "LZMAError\n",
      "RegistryError\n",
      "_GiveupOnFastCopy\n",
      "Error\n",
      "  |- NoSectionError\n",
      "  |- DuplicateSectionError\n",
      "  |- DuplicateOptionError\n",
      "  |- NoOptionError\n",
      "  |- InterpolationError\n",
      "  |- ParsingError\n",
      "NoIPAddresses\n",
      "BadZipFile\n",
      "LargeZipFile\n",
      "BadEntryPoint\n",
      "NoSuchEntryPoint\n",
      "DuplicateKernelError\n",
      "ErrorDuringImport\n",
      "NotOneValueFound\n",
      "CannotEval\n",
      "OptionError\n",
      "BdbQuit\n",
      "Restart\n",
      "ExceptionPexpect\n",
      "  |- EOF\n",
      "  |- TIMEOUT\n",
      "PtyProcessError\n",
      "FindCmdError\n",
      "HomeDirError\n",
      "ProfileDirError\n",
      "IPythonCoreError\n",
      "  |- TryNext\n",
      "  |- UsageError\n",
      "  |- StdinNotImplementedError\n",
      "InputRejected\n",
      "GetoptError\n",
      "ErrorToken\n",
      "PrefilterError\n",
      "AliasError\n",
      "  |- InvalidAliasError\n",
      "Error\n",
      "  |- InterfaceError\n",
      "  |- DatabaseError\n",
      "Warning\n",
      "SpaceInInput\n",
      "DOMException\n",
      "  |- IndexSizeErr\n",
      "  |- DomstringSizeErr\n",
      "  |- HierarchyRequestErr\n",
      "  |- WrongDocumentErr\n",
      "  |- InvalidCharacterErr\n",
      "  |- NoDataAllowedErr\n",
      "  |- NoModificationAllowedErr\n",
      "  |- NotFoundErr\n",
      "  |- NotSupportedErr\n",
      "  |- InuseAttributeErr\n",
      "  |- InvalidStateErr\n",
      "  |- SyntaxErr\n",
      "  |- InvalidModificationErr\n",
      "  |- NamespaceErr\n",
      "  |- InvalidAccessErr\n",
      "  |- ValidationErr\n",
      "ValidationError\n",
      "EditReadOnlyBuffer\n",
      "_Retry\n",
      "InvalidLayoutError\n",
      "HeightIsUnknownError\n",
      "ParserSyntaxError\n",
      "InternalParseError\n",
      "_PositionUpdatingFinished\n",
      "SimpleGetItemNotFound\n",
      "UncaughtAttributeError\n",
      "HasNoContext\n",
      "ParamIssue\n",
      "_JediError\n",
      "  |- InternalError\n",
      "  |- WrongVersion\n",
      "  |- RefactoringError\n",
      "OnErrorLeaf\n",
      "InvalidPythonEnvironment\n",
      "MessageError\n",
      "  |- MessageParseError\n",
      "  |- MultipartConversionError\n",
      "  |- CharsetError\n",
      "Error\n",
      "HTTPException\n",
      "  |- NotConnected\n",
      "  |- InvalidURL\n",
      "  |- UnknownProtocol\n",
      "  |- UnknownTransferEncoding\n",
      "  |- UnimplementedFileMode\n",
      "  |- IncompleteRead\n",
      "  |- ImproperConnectionState\n",
      "  |- BadStatusLine\n",
      "  |- LineTooLong\n",
      "InteractivelyDefined\n",
      "KillEmbedded\n",
      "Error\n",
      "  |- NoSuchProcess\n",
      "  |- AccessDenied\n",
      "  |- TimeoutExpired\n",
      "_Ipv6UnsupportedError\n",
      "QueueEmpty\n",
      "QueueFull\n",
      "DebuggerInitializationError\n",
      "ExpatError\n",
      "Error\n",
      "  |- ProtocolError\n",
      "  |- ResponseError\n",
      "  |- Fault\n",
      "ParseBaseException\n",
      "  |- ParseException\n",
      "  |- ParseFatalException\n",
      "RecursiveGrammarException\n",
      "ResolutionError\n",
      "  |- VersionConflict\n",
      "  |- DistributionNotFound\n",
      "  |- UnknownExtra\n",
      "_Error\n",
      "UnableToResolveVariableException\n",
      "InvalidTypeInArgsException\n"
     ]
    }
   ],
   "source": [
    "# iterate over the base Exception class and its subclasses to print the Python Exception Hierarchy\n",
    "for exc in Exception.__subclasses__():\n",
    "    print(exc.__name__)\n",
    "    for sub_exc in exc.__subclasses__():\n",
    "        print(\"  |-\", sub_exc.__name__)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bda84c0d-28b4-4c30-aa18-c26b6c124ff4",
   "metadata": {},
   "source": [
    "Q3. What errors are defined in the ArithmeticError class? Explain any two with an example.\n",
    "\n",
    "The ArithmeticError class is a built-in exception class in Python that serves as the base class for all exceptions that occur during arithmetic operations. It is a subclass of the Exception class.\n",
    "\n",
    "The errors defined in the ArithmeticError class are:\n",
    "\n",
    "FloatingPointError: Raised when a floating point operation fails.\n",
    "OverflowError: Raised when the result of an arithmetic operation is too large to be represented.\n",
    "ZeroDivisionError: Raised when trying to divide by zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "381732b9-fa9b-4ee8-ada4-155ba9d877d4",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "math domain error",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mmath\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m----> 6\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mmath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msqrt\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mFloatingPointError\u001b[39;00m:\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError: Cannot take the square root of a negative number.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mValueError\u001b[0m: math domain error"
     ]
    }
   ],
   "source": [
    "#FloatingPointError:\n",
    "\n",
    "import math\n",
    "\n",
    "try:\n",
    "    result = math.sqrt(-1)\n",
    "except FloatingPointError:\n",
    "    print(\"Error: Cannot take the square root of a negative number.\")\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "71394257-982e-4db3-b8ac-3f82b43f54a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#OverflowError\n",
    "import sys\n",
    "\n",
    "try:\n",
    "    x = sys.maxsize\n",
    "    result = x * x\n",
    "except OverflowError:\n",
    "    print(\"Error: Result of arithmetic operation is too large to be represented.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5907eca0-55d4-4261-9b16-75bfd0dfda48",
   "metadata": {},
   "source": [
    "Q4. Why LookupError class is used? Explain with an example KeyError and IndexError.\n",
    "\n",
    "The LookupError class is a built-in exception class in Python that serves as the base class for all exceptions that occur when looking up a value in a sequence or mapping. It is a subclass of the Exception class.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ac19dba5-e2a1-4d18-ada9-ef54a837c6ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Key not found in dictionary.\n"
     ]
    }
   ],
   "source": [
    "# KeyError: This error is raised when attempting to access a non-existent key in a dictionary or other mapping.\n",
    "\n",
    "my_dict = {\"apple\": 1, \"banana\": 2, \"orange\": 3}\n",
    "\n",
    "try:\n",
    "    value = my_dict[\"grape\"]\n",
    "except KeyError:\n",
    "    print(\"Error: Key not found in dictionary.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c0938204-080f-4ff4-a721-8d761a3fd56b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: Index out of range.\n"
     ]
    }
   ],
   "source": [
    "#IndexError: This error is raised when attempting to access a non-existent index in a sequence such as a list, tuple or string.\n",
    "\n",
    "my_list = [1, 2, 3]\n",
    "\n",
    "try:\n",
    "    value = my_list[3]\n",
    "except IndexError:\n",
    "    print(\"Error: Index out of range.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b8bb1bb-a2a0-4600-b566-ee2c832af37d",
   "metadata": {},
   "source": [
    "Q5. Explain ImportError. What is ModuleNotFoundError?\n",
    "\n",
    "\n",
    "ImportError and ModuleNotFoundError are both built-in exception classes in Python that are raised when there is an error while importing a module.\n",
    "\n",
    "ImportError is a base class for all import-related errors. It is raised when an imported module cannot be found, when a module is found but it has errors in it, or when a module is found but it cannot be loaded for some other reason.\n",
    "\n",
    "ModuleNotFoundError is a subclass of ImportError that was introduced in Python 3.6. It is raised when a module cannot be found during import."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a363855-0308-4b33-82aa-52c44f4838e9",
   "metadata": {},
   "source": [
    "Q6. List down some best practices for exception handling in python.\n",
    "\n",
    "Here are some best practices for exception handling in Python:\n",
    "\n",
    "1.Use specific exception types: Always use specific exception types instead of general ones. This makes the code easier to read and maintain. For example, use ValueError for invalid arguments, IOError for file-related errors, etc.\n",
    "\n",
    "2.Keep exception handling separate from business logic: Don't mix exception handling with business logic. Keep them separate so that the code is easier to understand.\n",
    "\n",
    "3.Handle exceptions at the right level: Handle exceptions at the right level of abstraction. Don't catch exceptions at a lower level than necessary. This makes it easier to identify the source of the problem.\n",
    "\n",
    "4.Use try/except blocks sparingly: Use try/except blocks only where necessary. Don't use them to control the flow of the program. Use if/else blocks instead.\n",
    "\n",
    "5.Use finally blocks: Always use finally blocks to perform cleanup operations, such as closing files or releasing resources.\n",
    "\n",
    "6.Don't catch all exceptions: Avoid catching all exceptions with a broad try statement. This can mask bugs and make it difficult to debug the code. Instead, catch only the specific exceptions you need to handle.\n",
    "\n",
    "7.Log exceptions: Always log exceptions to help with debugging. Use a logging framework like logging to log exceptions and their stack traces."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e61dbde-2bd4-47c3-80eb-bed75f82e49e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
